text/0017-message/0017-message.md

```diff
@@ -0,0 +1,167 @@
+
+# Summary
+
+A message is the object nodes gossip around in the network. It always references two other messages that are known as `parents`. It is stored as a vertex on the tangle data structure that the nodes maintain.
+
+The messages will contain payloads. Some of them will be core payloads that will be processed by all nodes as part of the core protocol. Some of them will be community payloads that will enable to build new functionality on top of the tangle. Some payloads may have other nested payloads embedded inside.
+So upon parsing it is done layer by layer.
+
+# Motivation
+
+To better understand this layered design, consider the internet protocol for example. There is an Ethernet frame, that contains an IP payload. This in turn contains a TCP packet that encapsulates an HTTP payload. Each layer has a certain responsibility, once this responsibility is completed, we move on to the next layer.
+
+The same goes with how we parse messages. The outer layer of the message enables us to map the message to a vertex in the Tangle and perform some basic validation. The next layer may be a transaction that mutates the ledger state. The next layer may provide some extra functionality on the transactions to be used by applications.
+
+By making it possible to add and exchange payloads, we are creating an architecture that can be easily extended to accommodate future needs.
+
+# Detailed design
+
+### Data types
+
+The following are data types that we will use when we specify fields in the message and payloads.
+
+| Name   | Description   |
+| ------ | ------------- |
+| uint8  | An unsigned 8 bit integer encoded in Little Endian. |
+| uint16  | An unsigned 16 bit integer encoded in Little Endian. |
+| uint32  | An unsigned 32 bit integer encoded in Little Endian. |
+| uint64  | An unsigned 64 bit integer encoded in Little Endian. |
+| ByteArray[N] | A static size array of size N.   |
+| ByteArray | A dynamically sized array. A uint32 denotes its length.   |
+| string | A dynamically sized array of an UTF-8 encoded string. A uint16 denotes its length.   |
+
+
+### Message ID
+The message ID will be the `BLAKE2b-256` hash of the byte contents of the message. It should be used by the nodes to index the messages and by external APIs.
+
+
+### Message structure
+
+<table>
+    <tr>
+        <th>Name</th>
+        <th>Type</th>
+        <th>Description</th>
+    </tr>
+    <tr>
+        <td>Version</td>
+        <td>uint8</td>
+        <td>The message version. The schema specified in this RFC is for version <strong>1</strong> only. </td>
+    </tr>
+    <tr>
+        <td>Parent1 (<code>trunk</code>)</td>
+        <td>ByteArray[32]</td>
+        <td>The Message ID of the first <i>Message</i> it references.</td>
+    </tr>
+    <tr>
+        <td>Parent2 (<code>branch</code>)</td>
+        <td>ByteArray[32]</td>
+        <td>The Message ID of the second <i>Message</i> it references.</td>
+    </tr>
+    <tr>
+        <td>Payload Length</td>
```

Comment:
is it memory length or serialization length of the payload?

Reply:
serialization length

After deserializing the payload, in theory a node implementation can choose any way it wants to represent it in memory.
We attempt to spec the protocol, not the implementation details (unless they are necessary)